src/vehicles/RosVehicle.lua

```diff
@@ -0,0 +1,257 @@
+limitations under the License.
+author: Ting-Chia Chiang
+author: G.A. vd. Hoorn
+--]]
+-- source(Utils.getFilename("src/modROS.lua", g_modROSModDirectory))
+
+RosVehicle = {}
+
+RosVehicle.MOD_DIR = g_modROSModDirectory
+RosVehicle.MOD_NAME = g_modROSModName
+
+
+function RosVehicle.prerequisitesPresent()
+    return true
+end
+
+
+function RosVehicle.initSpecialization()
+end
+
+
+function RosVehicle.registerFunctions(vehicleType)
+    SpecializationUtil.registerFunction(vehicleType, "getLaserFrameNode", RosVehicle.getLaserFrameNode)
+    SpecializationUtil.registerFunction(vehicleType, "pubOdom", RosVehicle.pubOdom)
+    SpecializationUtil.registerFunction(vehicleType, "addTF", RosVehicle.addTF)
+    SpecializationUtil.registerFunction(vehicleType, "fillLaserData", RosVehicle.fillLaserData)
+end
+
+
+function RosVehicle.registerEventListeners(vehicleType)
+    SpecializationUtil.registerEventListener(vehicleType, "onLoad", RosVehicle)
+    -- SpecializationUtil.registerEventListener(vehicleType, "onUpdate", RosVehicle)
+end
+
+
+-- -- rosVehicle spec own update
+-- function RosVehicle:onUpdate(dt)
+
+-- end
+
+
+function RosVehicle:onLoad()
+    -- rosVehicle namespace
+    self.spec_rosVehicle = self["spec_" .. RosVehicle.MOD_NAME .. ".rosVehicle"]
+    local spec = self.spec_rosVehicle
+    -- rosVehicle variables
+    spec.ros_veh_name = ros.Names.sanatize(self:getFullName() .. "_" .. self.id)
+    spec.base_link_frame = spec.ros_veh_name .. "/base_link"
+
+end
+
+
+function RosVehicle:pubOdom(ros_time, tf_msg, pub_odom)
+
+    local spec = self.spec_rosVehicle
+
+    -- legalize a name for ROS
+    -- local vehicle_name = ros.Names.sanatize(vehicle:getFullName() .. "_" .. vehicle.id)
+    local vehicle_name = spec.base_link_frame
+
+
+    -- retrieve the vehicle node we're interested in
+    local veh_node = self.components[1].node
+
+    -- retrieve global (ie: world) coordinates of this node
+    local p_x, p_y, p_z = getWorldTranslation(veh_node)
+
+    -- retrieve global (ie: world) quaternion of this node
+    local q_x, q_y, q_z, q_w = getWorldQuaternion(veh_node)
+
+    -- get twist data
+    local l_v_x, l_v_y, l_v_z = getLocalLinearVelocity(veh_node)
+    -- we don't use getAngularVelocity(veh_node) here as the return value is wrt the world frame not local frame
+
+
+    -- convert y up world to z up world (farmsim coordinate system: x right, z towards me, y up; ROS: y right, x towards me, z up)
+    -- https://stackoverflow.com/questions/16099979/can-i-switch-x-y-z-in-a-quaternion
+    -- https://gamedev.stackexchange.com/questions/129204/switch-axes-and-handedness-of-a-quaternion
+    -- https://stackoverflow.com/questions/18818102/convert-quaternion-representing-rotation-from-one-coordinate-system-to-another
+
+    -- FS time is "frozen" within a single call to update(..), so this
+    -- will assign the same stamp to all Odometry messages
+    local t = ros_time
```

Comment:
I would suggest to use `ros_time` directly below when setting the stamp of the `Odometry` message and the TF message. There doesn't seem to be any advantage to aliasing it to this local variable here.